### PR TITLE
Add CPU fallback for BMP images and concious fail for GIF

### PR DESCRIPTION
**Pull request by JanuszL**
Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

**Comment by JanuszL** [comment]
I don't know if it is worth to deal with TIFFs as well. So let leave it for latter.

File: dali/image/png.cc [hunk 1]
```diff
@@ -36,6 +37,11 @@ DALIError_t GetPNGImageDims(const uint8 *png, int size, int *h, int *w) {
 
   // IHDR needs to be the first chunk
   const uint8 *IHDR = png + 8;
+  const uint8 *png_size = IHDR;
+  if (IHDR[4] != 'I' && IHDR[5] != 'H' && IHDR[6] != 'D' && IHDR[7] != 'R') {
```
Comment (cliffwoolley) [inline dali/image/png.cc]: should these be `||` instead of `&&`?

File: dali/image/generic_image.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    DALI_ASSERT(gif);
+    return (size > 9 && gif[0] == 'G' && gif[1] == 'I' && gif[2] == 'F' && gif[3] == '8' &&
+    (gif[4] == '7' || gif[4] == '9') && gif[5] == 'a');
+}
+
+DALIError_t GetGIFImageDims(const uint8 *gif, int size, int *h, int *w) {
+    DALI_ASSERT(gif);
+
+    *w = (unsigned int)(gif[6] | gif[7] << 4) & 0xFFFF;
+    *h = (unsigned int)(gif[8]  | gif[9]<< 4) & 0xFFFF;
```
Comment (cliffwoolley) [inline dali/image/generic_image.cc]: nit: whitespace consistency

File: dali/image/generic_image.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+DALIError_t GetImageDims(const uint8 *data, int size, int *h, int *w) {
+    DALI_ASSERT(data);
+    if (CheckIsPNG(data, size)) {
+        return GetPNGImageDims(data, size, h, w);
+    } else if (CheckIsGIF(data, size)) {
+        // OpenCV doesn't handle gif images
+        DALI_FAIL("GIF format is not supported");
+    #if 0
+        return GetGIFImageDims(data, size, h, w);
+    #endif
```
Comment (cliffwoolley) [inline dali/image/generic_image.cc]: Should the GetGIFImageDims() function definition itself also be `#if 0` ?

File: dali/image/png.cc [hunk 4]
```diff
@@ -36,6 +37,11 @@ DALIError_t GetPNGImageDims(const uint8 *png, int size, int *h, int *w) {
 
   // IHDR needs to be the first chunk
   const uint8 *IHDR = png + 8;
```
Comment (cliffwoolley) [inline dali/image/png.cc]: Should there be a bounds check on size>8?

File: dali/image/png.cc [hunk 5]
```diff
@@ -47,8 +53,8 @@ DALIError_t GetPNGImageDims(const uint8 *png, int size, int *h, int *w) {
   // 1 byte : Compression method
   // 1 byte : Filter method
   // 1 byte : Interlace method
-  *w = ReadIntFromPNG(IHDR + 8);
-  *h = ReadIntFromPNG(IHDR + 12);
+  *w = ReadIntFromPNG(png_size + 8);
+  *h = ReadIntFromPNG(png_size + 12);
```
Comment (cliffwoolley) [inline dali/image/png.cc]: Bounds checking?
Also maybe it's just me but `png_size` is a confusing name for what I think this is actually pointing to?

File: dali/image/generic_image.cc (hunk 3 above)
Comment (JanuszL) [reply dali/image/generic_image.cc]: Done

File: dali/image/png.cc (hunk 1 above)
Comment (JanuszL) [reply dali/image/png.cc]: Done

File: dali/image/generic_image.cc (hunk 2 above)
Comment (JanuszL) [reply dali/image/generic_image.cc]: Done

File: dali/image/png.cc (hunk 4 above)
Comment (JanuszL) [reply dali/image/png.cc]: Done

File: dali/image/png.cc (hunk 5 above)
Comment (JanuszL) [reply dali/image/png.cc]: Done

File: dali/image/generic_image.cc [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+// OpenCV doesn't handle gif images so we don't need it now
+#if 0
+DALIError_t GetGIFImageDims(const uint8 *gif, int size, int *h, int *w) {
+    DALIError_t ret = DALIError;
+    DALI_ASSERT(gif);
+
+    if (size >= 10) {
+        *w = (unsigned int)(gif[6] | gif[7] << 4) & 0xFFFF;
+        *h = (unsigned int)(gif[8] | gif[9]<< 4) & 0xFFFF;
+        ret = DALISuccess;
```
Comment (pribalta) [inline dali/image/generic_image.cc]: At this point might be worth checking that w*h is the same as size minus the length of the GIF header. Might help catching malformed images or incorrect allocations

File: dali/image/generic_image.cc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    DALIError_t ret = DALIError;
+    DALI_ASSERT(bmp);
+
+    unsigned head_size = bmp[14] | bmp[15] << 4 | bmp[16] << 8 | bmp[17] << 12;
+    *h = 0;
+    *w = 0;
+    if (size >= 22 && head_size == 12) {
+        *w = (unsigned int)(bmp[18] | bmp[19] << 4) && 0xFFFF;
+        *h = (unsigned int)(bmp[20] | bmp[21] << 4) && 0xFFFF;
+        ret = DALISuccess;
```
Comment (pribalta) [inline dali/image/generic_image.cc]: Same as line 37 but with BMP header

File: dali/image/generic_image.cc [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    } else if (size >= 26 && head_size >= 40) {
+        *w = (unsigned int)(static_cast<int>(bmp[18] | bmp[19] << 4 |
+             bmp[20] << 8 | bmp[21] << 12));
+        *h = (unsigned int)abs(static_cast<int>(bmp[22] | bmp[23] << 4 |
+             bmp[24] << 8 | bmp[25] << 12));
+        ret = DALISuccess;
```
Comment (pribalta) [inline dali/image/generic_image.cc]: Same as line 37 but with BMP header

File: dali/image/generic_image.cc (hunk 6 above)
Comment (JanuszL) [reply dali/image/generic_image.cc]: No really, as size parameter her is the size of the data we have read, and as GIF is compressed file, so w*h is suppose to be bigger than the file size.

File: dali/image/generic_image.cc (hunk 7 above)
Comment (JanuszL) [reply dali/image/generic_image.cc]: BMP format allows few different color pallets (16, 256 or 24bits for example) and calculating real image size (including padding) would complicate logic here. I would prefer to leave it for OpenCV.

File: dali/image/generic_image.cc (hunk 8 above)
Comment (JanuszL) [reply dali/image/generic_image.cc]: Same as above.

File: dali/image/generic_image.cc [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        // OpenCV doesn't handle gif images
+        DALI_FAIL("GIF format is not supported");
+    #if 0
+        return GetGIFImageDims(data, size, h, w);
+    #endif
+    } else if (CheckIsBMP(data, size)) {
+        return GetBMPImageDims(data, size, h, w);
+    }
+    // Not supported
+    return DALIError;
```
Comment (ptrendx) [inline dali/image/generic_image.cc]: I would prefer `DALI_FAIL("Unsupported image format - DALI supports JPEG, PNG and BMP formats.")` instead - having 2 different failure methods (via DALIError_t and exceptions) sucks.

Comment (JanuszL) [reply dali/image/generic_image.cc]: Done. I think I will leave this whole file to return only DALIError_t and make nvjpeg_decoder.h to throw something.

**Pull request merged**